### PR TITLE
SRC-9; Metadata Keys Standard

### LINKED ISSUE #26: SRC-9; Metadata Keys Standard
# Abstract

The following standard attempts to define the keys of relevant on-chain metadata for any [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets). Any contract that implements the SRC-8 standard MUST implement the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) and [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) standards. This is a living standard where revisions may be made as the ecosystem evolves.

# Motivation

The SRC-9 standard seeks to enable relevant data for assets on the Fuel Network. This data may include images, text, contact, or all of the above. All metadata queries are done through a single function to facilitate cross-contract calls.

# Prior Art

The use of generic metadata for [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets) is defined in the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard. This standard integrates into the existing [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard.

# Specification

The following keys are reserved for the SRC-9 standard. Use of the key should follow the SRC-9 specification.

All keys SHALL use snake case. 

## Social 

The social prefix SHALL be used for any social media platform and SHALL return usernames.

Any social media metadata keys SHALL follow the following syntax `social:site` where:
-  The `social` keyword must be prepended to denote this is a social platform
- The `site` keyword must be the website or platform of the social

#### - `social:discord`

The key `social:discord` SHALL return a `String` variant of a username for the discord platform.

#### - `social:facebook`

The key `social:facebook` SHALL return a `String` variant of a username for the Facebook platform.

#### - `social:farcaster`

The key `social:farcaster` SHALL return a `String` variant of a username for the Farcaster platform.

#### - `social:friend.tech`

The key `social:friend.tech` SHALL return a `String` variant of a username for the Friend.tech platform.

#### - `social:github`

The key `social:github` SHALL return a `String` variant of a username for the Github platform.

#### - `social:instagram`

The key `social:instagram` SHALL return a `String` variant of a username for the Instagram platform.

#### - `social:lens`

The key `social:lens` SHALL return a `String` variant of a username for the Lens Protocol.

#### - `social:linkedin`

The key `social:linkedin` SHALL return a `String` variant of a username for the LinkedIn platform.

#### - `social:reddit`

The key `social:reddit` SHALL return a `String` variant of a username for the Reddit platform.

#### - `social:signal`

The key `social:signal` SHALL return a `String` variant of a username for the Signal platform.

#### - `social:telegram`

The key `social:telegram` SHALL return a `String` variant of a username for the Telegram platform.

#### - `social:tiktok`

The key `social:tiktok` SHALL return a `String` variant of a username for the TikTok platform.

#### - `social:x`

The key `social:x` SHALL return a `String` variant of a username for the X or formerly Twitter platform.

#### - `social:wechat`

The key `social:wechat` SHALL return a `String` variant of a username for the WeChat platform.

#### - `social:whatsapp`

The key `social:whatsapp` SHALL return a `String` variant of a username for the WhatsApp platform.

#### - `social:youtube`

The key `social:youtube` SHALL return a `String` variant of a username for the YouTube platform.

## Contact

The `contact` prefix SHALL be used for any contact information on a particular project's team for an asset.

Any contact information metadata keys SHALL follow the following syntax `contract:type` where:
-  The `contact` keyword must be prepended to denote this is contact information
- The `type` keyword must be the method of contact

The key SHALL use snake case. 

#### - `contact:email`

The key `contact:email` SHALL return a `String` variant of an email.

#### - `contact:mailing`

The key `contact:mailing` SHALL return a `String` variant of a mailing address. All mailing addresses MUST follow the UPU addressing format.

#### - `contact:phone`

The key `contact:phone` SHALL return a `String` variant of a phone number. All phone numbers SHALL follow the E.164 standard.

#### - `contact:company`

The key `contact:company` SHALL return a `String` variant of a company name.

## External Links

The `link` prefix SHALL be used for any external webpage hyperlink associated with an asset.

Any external webpage metadata keys SHALL follow the following syntax `link:site` where:
-  The `link` keyword must be prepended to denote this is an external webpage
- The `site` keyword must be an external website

#### - `link:home`

The key `link:home` SHALL return a `String` variant of the asset's project homepage.

#### - `link:contact`

The key `link:contact` SHALL return a `String` variant of the asset's project contact information webpage.

#### - `link:docs`

 The key `link:docs` SHALL return a `String` variant of the asset's project documentation webpage.

#### - `link:forum`

The key `link:forum` SHALL return a `String` variant of the asset's project forum webpage.

#### - `link:blog`

The key `link:blog` SHALL return a `String` variant of the asset's project blog.

#### - `link:linktree`

The key `link:linktree` SHALL return a `String` variant of the asset's project linktree information webpage.

## Resources

The `res` prefix SHALL be used for any resources or general information on an asset.

Any resource metadata keys SHALL follow the following syntax `rec:type` where:
-  The `res` keyword must be prepended to denote this is a resource
- The `type` keyword must be the type of resource

#### - `res:license`

The key `res:license` SHALL return a `String` variant of the asset's project license.

#### - `res:tos`

The key `res:tos` SHALL return a `String` variant of the asset's project Terms of Service.

#### - `res:author`

The key `res:author` SHALL return a `String` variant of the asset's project author. This MAY be a full name or pseudonym.

#### - `res:about`

The key `res:about` SHALL return a `String` variant about the asset's project up to 2048 characters.

#### - `res:description`

The key `res:description` SHALL return a `String` variant describing the asset's project up to 256 characters.

#### - `res:date`

The key `res:date` SHALL return a `Int` variant of a UNIX timestamp.

#### - `res:block`

The key `res:block` SHALL return a `Int` variant of a block number.

## Images

The `image` prefix SHALL be used for any image files associated with a singular token.

Any image metadata keys SHALL follow the following syntax `image:type` where:
-  The `image` keyword must be prepended to denote this is an image
- The `type` keyword must be the file type of the image

#### - `image:svg`

The key `image:svg` SHALL return a `String` variant of an SVG image.

#### - `image:png`

The key `image:png` SHALL return a `String` variant of a URI for a PNG image.

#### - `image:jpeg`

The key `image:jpeg` SHALL return a `String` variant of a URI for a JPEG image.

#### - `image:webp`

The key `image:webp` SHALL return a `String` variant of a URI for a WebP image.

#### - `image:gif`

The key `image:gif` SHALL return a `String` variant of a URI for a GIF image.

#### - `image:heif`

The key `image:heif` SHALL return a `String` variant of a URI for a HEIF image.

## Video

The `video` prefix SHALL be used for any video files associated with a singular token.

Any video metadata keys SHALL follow the following syntax `video:type` where:
-  The `video` keyword must be prepended to denote this is a video
- The `type` keyword must be the file type of the video

#### - `video:mp4`

The key `video:mp4` SHALL return a `String` variant of a URI for an MP4 video.

#### - `video:webm`

The key `video:webm` SHALL return a `String` variant of a URI for a WebM video.

#### - `video:m4v`

The key `video:m4v` SHALL return a `String` variant of a URI for a M4V video.

#### - `video:ogv`

The key `video:ogv` SHALL return a `String` variant of a URI for an OGV video.

#### - `video:ogg`

The key `video:ogg` SHALL return a `String` variant of a URI for an OGG video.

## Audio

The `audio` prefix SHALL be used for any audio files associated with a singular token.

Any audio metadata keys SHALL follow the following syntax `audio:type` where:
-  The `audio` keyword must be prepended to denote this is audio metadata
- The `type` keyword must be the file type of the audio

#### - `audio:mp3`

The key `audio:mp3` SHALL return a `String` variant of a URI for an MP3 file.

#### - `audio:wav`

The key `audio:wav` SHALL return a `String` variant of a URI for a WAV file.

#### - `audio:oga`

The key `audio:oga` SHALL return a `String` variant of a URI for an OGA file.

## Media

The `media` prefix SHALL be used for any media associated with a particular singular token.

Any media metadata keys SHALL follow the following syntax `media:type` where:
-  The `media` keyword must be prepended to denote this is a video
- The `type` keyword must be the file type of the media

#### - `media:gltf`

The key `media:gltf` SHALL return a `String` variant of a URI for a glTF file.

#### - `media:glb`

The key `media:glb` SHALL return a `String` variant of a URI for a GLB file.

## Logos

The `logo` prefix SHALL be used for any images associated with a particular asset or project.

Any logo metadata keys SHALL follow the following syntax `logo:type` where:
-  The `logo` keyword must be prepended to denote this is a logo
- The `type` keyword must be the type of logo

#### - `logo:svg_light`

The key `logo:svg_light` SHALL return a `String` variant of an SVG image of a logo for light themes.

#### - `logo:svg_dark`

The key `logo:svg_light` SHALL return a `String` variant of an SVG image of a logo for dark themes.

#### - `logo:small_light`

The key `logo:small_light` SHALL return a `String` variant of a URI for a 32x32 PNG image of a logo for light themes.

#### - `logo:small_dark`

The key `logo:small_dark` SHALL return a `String` variant of a URI for a 32x32 PNG image of a logo for dark themes.

#### - `logo:medium_light`

The key `logo:medium_light` SHALL return a `String` variant of a URI for a 256x256 PNG image of a logo for light themes.

#### - `logo:medium_dark`

The key `logo:medium_dark` SHALL return a `String` variant of a URI for a 256x256 PNG image of a logo for dark themes.

#### - `logo:large_light`

The key `logo:large_light` SHALL return a `String` variant of a URI for a 1024x1024 PNG image of a logo for light themes.

#### - `logo:large_dark`

The key `logo:large_dark` SHALL return a `String` variant of a URI for a 1024x1024 PNG image of a logo for dark themes.

## Attributes

The `attr` prefix SHALL be used for any attributes associated with a singular token.

Any attribute metadata keys SHALL follow the following syntax `attr:type` where:
-  The `attr` keyword must be prepended to denote this is an attribute
- The `type` keyword must be the type of attribute

There are no standardized types of attributes.
Example: `attr:eyes`.

# Rationale

The SRC-9 standard should allow for standardized keys for metadata on the Fuel Network. This standard builds off existing standards and should allow other contracts to query any relevant information on the asset.

# Backwards Compatibility

This standard is compatible with Fuel's [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets), the [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) standard, and the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard. 

# Security Considerations

This standard does not call external contracts, nor does it define any mutations of the contract state.

# Example

```rust
impl SRC7 for Contract {
    fn metadata(asset: AssetId, key: String) -> Option<Metadata> {
        if (asset != AssetId::from(ZERO_B256)) {
            return Option::None;
        }

        match key {
            String::from_ascii_str("social:x") => {
                let social = String::from_ascii_str("fuel_network");
                Option::Some(Metadata::String(social))
            },
            _ => Option::None,
        }
    }
}
```

### PR DESCRIPTION
**Pull request by bitzoic**
## Type of change

- New standard

## Changes

The following changes have been made:

- Defines the SRC-9 standard, creating a standardized list of keys for different types of metadata

## Notes

- This is a living standard and can be expected to get additions as the ecosystem evolves

## Related Issues

Closes #26

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ If you don't find what you're looking for, feel free to create an issue and prop
 - [SRC-3; Mint and Burn](./standards/src_3/) is used to enable mint and burn functionality for Native Assets.
 - [SRC-5; Ownership Standard](./standards/src_5/) is used to restrict function calls to admin users in contracts.
 - [SRC-7; Arbitrary Asset Metadata Standard](./standards/src_7/) is used to store metadata for [Native Assets](https://fuellabs.github.io/sway/v0.44.0/book/blockchain-development/native_assets.html).
+- [SRC-9; Metadata Keys Standard](./standards/src_9/) is used to store standardize metdata keys for [Native Assets](https://fuellabs.github.io/sway/v0.44.0/book/blockchain-development/native_assets.html) in combination with the SRC-7 standard.
```
Comment (Braqzen) [inline README.md:41]:
```suggestion
- [SRC-9; Metadata Keys Standard](./standards/src_9/) is used to store standardized metadata keys for [Native Assets](https://fuellabs.github.io/sway/v0.44.0/book/blockchain-development/native_assets.html) in combination with the SRC-7 standard.
```

File: standards/src_9/README.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    </picture>
+</p>
+
+# Abstract
+
+The following standard attempts to define the keys of relevant on-chain metadata for any [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets). Any contract that implements the SRC-8 standard MUST implement the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) and [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) standards. This is a living standard where revisions may be made as the ecosystem evolves.
+
+# Motivation
+
+The SRC-9 standard seeks to enable relevant data for assets on the Fuel Network. This data may include images, text, contact, or all of the above. All metadata queries are done through a single function to facilitate cross-contract calls.
```
Comment (Braqzen) [inline standards/src_9/README.md:14]: > The SRC-9 standard seeks to enable relevant data for assets on the Fuel Network.

This doesn't really make sense to me. "To enable relevant data". It seems as if it's missing a piece of information. How does it enable it? How is it currently not "enabled"?

File: standards/src_9/README.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+# Prior Art
+
+The use of generic metadata for [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets) is defined in the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard. This standard integrates into the existing [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard.
+
+# Specification
+
+The following keys are reserved for the SRC-9 standard. Use of the key should follow the SRC-9 specification.
```
Comment (Braqzen) [inline standards/src_9/README.md:22]:
```suggestion
The following keys are reserved for the SRC-9 standard. Use of the keys should follow the SRC-9 specification.
```

File: standards/src_9/README.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+The social prefix SHALL be used for any social media platform and SHALL return usernames.
+
+Any social media metadata keys SHALL follow the following syntax `social:site` where:
+-  The `social` keyword must be prepended to denote this is a social platform
+- The `site` keyword must be the website or platform of the social
+
+#### - `social:discord`
+
+The key `social:discord` SHALL return a `String` variant of a username for the discord platform.
```
Comment (Braqzen) [inline standards/src_9/README.md:36]:
```suggestion
The key `social:discord` SHALL return a `String` variant of a username for the Discord platform.
```

File: standards/src_9/README.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+# Motivation
+
+The SRC-9 standard seeks to enable relevant data for assets on the Fuel Network. This data may include images, text, contact, or all of the above. All metadata queries are done through a single function to facilitate cross-contract calls.
+
+# Prior Art
+
+The use of generic metadata for [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets) is defined in the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard. This standard integrates into the existing [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) standard.
+
+# Specification
```
Comment (Braqzen) [inline standards/src_9/README.md:20]: There's a lot of examples in the sections below. I would limit it to at most 4 most common / know per section and somewhere add a comment to indicate that it's not limited to just those 4. Easier to maintain and read. No need to scroll as much - lower cognitive load, cleaner.

File: standards/src_9/README.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+#### - `contact:company`
+
+The key `contact:company` SHALL return a `String` variant of a company name.
+
+## External Links
+
+The `link` prefix SHALL be used for any external webpage hyperlink associated with an asset.
+
+Any external webpage metadata keys SHALL follow the following syntax `link:site` where:
+-  The `link` keyword must be prepended to denote this is an external webpage
```
Comment (Braqzen) [inline standards/src_9/README.md:129]:
```suggestion
- The `link` keyword must be prepended to denote this is an external webpage
```

File: standards/src_9/README.md [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+#### - `contact:phone`
+
+The key `contact:phone` SHALL return a `String` variant of a phone number. All phone numbers SHALL follow the E.164 standard.
+
+#### - `contact:company`
+
+The key `contact:company` SHALL return a `String` variant of a company name.
+
+## External Links
```
Comment (Braqzen) [inline standards/src_9/README.md:124]: I believe links may also contain internal file system paths. Should we consider these as well or links to webpages is sufficient?

File: standards/src_9/README.md [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+#### - `social:youtube`
+
+The key `social:youtube` SHALL return a `String` variant of a username for the YouTube platform.
+
+## Contact
+
+The `contact` prefix SHALL be used for any contact information on a particular project's team for an asset.
+
+Any contact information metadata keys SHALL follow the following syntax `contract:type` where:
+-  The `contact` keyword must be prepended to denote this is contact information
```
Comment (Braqzen) [inline standards/src_9/README.md:103]:
```suggestion
- The `contact` keyword must be prepended to denote this is contact information
```

File: standards/src_9/README.md [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+The following keys are reserved for the SRC-9 standard. Use of the key should follow the SRC-9 specification.
+
+All keys SHALL use snake case. 
+
+## Social 
+
+The social prefix SHALL be used for any social media platform and SHALL return usernames.
+
+Any social media metadata keys SHALL follow the following syntax `social:site` where:
+-  The `social` keyword must be prepended to denote this is a social platform
```
Comment (Braqzen) [inline standards/src_9/README.md:31]:
```suggestion
- The `social` keyword must be prepended to denote this is a social platform
```

File: standards/src_9/README.md [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+#### - `image:heif`
+
+The key `image:heif` SHALL return a `String` variant of a URI for a HEIF image.
+
+## Video
+
+The `video` prefix SHALL be used for any video files associated with a singular token.
+
+Any video metadata keys SHALL follow the following syntax `video:type` where:
+-  The `video` keyword must be prepended to denote this is a video
```
Comment (Braqzen) [inline standards/src_9/README.md:229]:
```suggestion
- The `video` keyword must be prepended to denote this is a video
```

File: standards/src_9/README.md [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+#### - `video:ogg`
+
+The key `video:ogg` SHALL return a `String` variant of a URI for an OGG video.
+
+## Audio
+
+The `audio` prefix SHALL be used for any audio files associated with a singular token.
+
+Any audio metadata keys SHALL follow the following syntax `audio:type` where:
+-  The `audio` keyword must be prepended to denote this is audio metadata
```
Comment (Braqzen) [inline standards/src_9/README.md:257]:
```suggestion
- The `audio` keyword must be prepended to denote this is audio metadata
```

File: standards/src_9/README.md [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+#### - `audio:oga`
+
+The key `audio:oga` SHALL return a `String` variant of a URI for an OGA file.
+
+## Media
+
+The `media` prefix SHALL be used for any media associated with a particular singular token.
+
+Any media metadata keys SHALL follow the following syntax `media:type` where:
+-  The `media` keyword must be prepended to denote this is a video
```
Comment (Braqzen) [inline standards/src_9/README.md:277]:
```suggestion
- The `media` keyword must be prepended to denote this is a video
```

File: standards/src_9/README.md [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+#### - `media:glb`
+
+The key `media:glb` SHALL return a `String` variant of a URI for a GLB file.
+
+## Logos
+
+The `logo` prefix SHALL be used for any images associated with a particular asset or project.
+
+Any logo metadata keys SHALL follow the following syntax `logo:type` where:
+-  The `logo` keyword must be prepended to denote this is a logo
```
Comment (Braqzen) [inline standards/src_9/README.md:293]:
```suggestion
- The `logo` keyword must be prepended to denote this is a logo
```

File: standards/src_9/README.md [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+- The `type` keyword must be the type of logo
+
+#### - `logo:svg_light`
+
+The key `logo:svg_light` SHALL return a `String` variant of an SVG image of a logo for light themes.
+
+#### - `logo:svg_dark`
+
+The key `logo:svg_light` SHALL return a `String` variant of an SVG image of a logo for dark themes.
```
Comment (Braqzen) [inline standards/src_9/README.md:302]:
```suggestion
The key `logo:svg_dark` SHALL return a `String` variant of an SVG image of a logo for dark themes.
```

File: standards/src_9/README.md [hunk 15]
```diff
@@ -0,0 +1,369 @@
+<p align="center">
+    <picture>
+        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-9-logo-dark-theme.png">
+        <img alt="SRC-9 logo" width="400px" src=".docs/src-9-logo-light-theme.png">
+    </picture>
+</p>
+
+# Abstract
+
+The following standard attempts to define the keys of relevant on-chain metadata for any [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets). Any contract that implements the SRC-8 standard MUST implement the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) and [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) standards. This is a living standard where revisions may be made as the ecosystem evolves.
```
Comment (Braqzen) [inline standards/src_9/README.md:10]:
```suggestion
The following standard attempts to define the keys of relevant on-chain metadata for any [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets). Any contract that implements the SRC-9 standard MUST implement the [SRC-7](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_7) and [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) standards. This is a living standard where revisions may be made as the ecosystem evolves.
```

File: standards/src_9/README.md (hunk 5 above)
Comment (bitzoic) [reply standards/src_9/README.md:20]: No, this standard has been created with the intent to outline every reserved key without leaving any guesswork.

File: standards/src_9/README.md (hunk 7 above)
Comment (bitzoic) [reply standards/src_9/README.md:124]: Internal paths make no sense on a contract deployed on a blockchain. Internal to where?

File: standards/src_9/README.md (hunk 5 above)
Comment (Braqzen) [reply standards/src_9/README.md:20]: Okay, but the obvious response is that there are a lot more, say, websites that can be linked. If you are saying that these are exhausting then you are standardizing away a lot of ways that others may want to store their data.

I stand by my approach to limit and add a comment

File: standards/src_9/README.md (hunk 7 above)
Comment (Braqzen) [reply standards/src_9/README.md:124]: Isn't this data stored on chain? Doesn't it stand to reason that someone may be inclined to create some internal registry (even though this case may be expensive even in the Fuel network).

Similarly, the integration of servers and blockchain is a viable case. The chain ought to be used more as a trusted open database and so we could fetch the data and do something in the server.

Just a thought. I'm not saying we should care here, only asking if it has been considered.

File: standards/src_9/README.md [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+
+## Logos
+
+The `logo` prefix SHALL be used for any images associated with a particular asset or project.
+
+Any logo metadata keys SHALL follow the following syntax `logo:type` where:
+- The `logo` keyword must be prepended to denote this is a logo
+- The `type` keyword must be the type of logo
+
+#### - `logo:svg_light`
```
Comment (dmihal) [inline standards/src_9/README.md:296]: What will be the default if a logo is neither light nor dark? Does `logo:svg` make sense as well?

**Review by dmihal — APPROVED** [review]

**Review by SwayStar123 — APPROVED** [review]

**Pull request merged**